File: Javascript/DS/LinkedList.js
class Node {
    constructor(data, next = null) {
        this.data = data
        this.next = next
    }
}

class LinkedList {
    constructor() {
        this.head = null
        this.size = 0
    }

    // Insert First Node
    insertFirst(data) {
        this.head = new Node(data, this.head)
        this.size++
    }

    // Insert Last Node
    insertLast(data) {
        let node = new Node(data)
        let current

        // If Empty then make it head
        if(!this.head) {
            this.head = node
        } else {
            current = this.head
            while(current.next) {
                current = current.next
            }
            current.next = node
        }

        this.size++
    }


    // Insert at Index
    insertAt(data, index) {
        // if index out of range
        if(index > 0 && index > this.size) {
            return
        }

        // if first index
        if(index === 0) {
            this.insertFirst(data)
            return
        }

        const node = new Node(data)
        let current, previous

        // Set current to first
        current = this.head
        let count = 0
        
        while(count < index) {
            previous = current // node before index
            count++
            current = current.next // node after index
        }

        node.next = current
        previous.next = node

        this.size++
    }

    // Get at Index
    getAt(index) {
        let current = this.head
        let count = 0

        while(current) {
            if(count == index) {
                console.log(current.data)
                return
            }
            count++
            current = current.next
        }

        return null
    }


    // Remove at Index
    removeAt(index) {
        if(index > 0 && index > this.size) {
            return 
        }

        let current = this.head
        let previous
        let count = 0

        if(index === 0) {
            this.head = current.next
        } else {
            while(count < index) {
                count++
                previous = current
                current = current.next
            }
            previous.next = current.next
        }

        this.size--
    }

    // Clear List
    clearList() {
        this.head = null
        this.size = 0
    }


    // Print List Data
    printListData() {
        let current = this.head
        while(current) {
            console.log(current.data)
            current = current.next
        }
    }
}

const link = new LinkedList()
link.insertFirst(100)
link.insertFirst(200)
link.insertFirst(300)
link.insertLast(400)
link.insertAt(500, 2)
link.insertAt(800, 0)
link.insertAt(900, 10)  // test case because 10 is not in length


// link.printListData()

// console.log("Get by Index")
// link.getAt(3)

// link.removeAt(3)
// link.printListData()

link.clearList()
link.printListData()

File: Python/dsa/LinkedList.py
class Node:
    def __init__(self, data=None, next=None):
        self.data = data
        self.next = next

class LinkedList:
    def __init__(self):
        self.head = None
        self.__count = 0

    def insert_at_begining(self, data):
        node = Node(data, self.head)
        self.head = node
        self.__count += 1

    def print_list(self):
        if self.head is None:
            print('Empty List')
            return
        
        itr = self.head
        while itr:
            print(itr.data)
            itr = itr.next

    def insert_at_last(self, data):
        self.__count += 1
        if(self.head is None):
            self.head = Node(data, None)
            return
        itr = self.head
        while itr.next:
            itr = itr.next
        
        itr.next = Node(data, None)

    def get_count(self):
        return self.__count

    def remove_at(self, index):
        if index < 0 or index >= self.__count:
            raise Exception('Invalid Index')

        if index == 0:
            self.head = self.head.next
            return
        
        count = 0
        itr = self.head

        while itr:
            if count == index - 1:
                itr.next = itr.next.next
                return
            itr = itr.next
            count += 1

    def insert_at(self, data, index):
        if index < 0 or index >= self.__count:
            raise Exception('Invalid Index')

        if index == 0:
            self.insert_at_begining(data)
            return

        count = 0
        itr = self.head
        while itr:
            if count == index - 1:
                node = Node(data, itr.next)
                itr.next = node
                break
            itr = itr.next
            count += 1
        

if __name__ == '__main__':
    ll = LinkedList()
    ll.insert_at_begining(10)
    ll.insert_at_begining(20)
    ll.insert_at_begining(30)
    ll.insert_at_last(40)

    ll.print_list()
    print("Count of LinkedList is: ", ll.get_count())


    # ll.remove_at(10)
    ll.remove_at(2)
    ll.print_list()

    print("\n")

    ll.insert_at(50, 0)
    ll.print_list()

    print("\n")

    ll.insert_at(60, 2)
    ll.print_list()